src/configs/config.ts

```diff
@@ -0,0 +1,217 @@
+import { TimeDelta } from '../types';
+import {
+  API_ENDPOINT,
+  AutoblocksEnvVar,
+  readEnv,
+  AUTOBLOCKS_HEADERS,
+  convertTimeDeltaToMilliSeconds,
+} from '../util';
+import {
+  Config,
+  ConfigParameter,
+  ConfigSpecialVersion,
+  ConfigVersion,
+  zConfigSchema,
+} from './types';
+
+/**
+ * Note that we check for the presence of the CLI environment
+ * variable and not the test case local storage because the
+ * local storage vars aren't set until runTestSuite is called,
+ * whereas a AutoblocksConfig might have already been imported
+ * and activated by the time runTestSuite is called.
+ */
+const isTestingContext = (): boolean => {
+  return readEnv(AutoblocksEnvVar.AUTOBLOCKS_CLI_SERVER_ADDRESS) !== undefined;
+};
+
+/**
+ * The AUTOBLOCKS_CONFIG_REVISIONS environment variable is a JSON-stringified
+ * map of config IDs to revision IDs. This is set in CI test runs triggered
+ * from the UI.
+ */
+const configRevisionsMap = (): Record<string, string> => {
+  if (!isTestingContext()) {
+    return {};
+  }
+
+  const configRevisionsRaw = readEnv(
+    AutoblocksEnvVar.AUTOBLOCKS_CONFIG_REVISIONS,
+  );
+  if (!configRevisionsRaw) {
+    return {};
+  }
+
+  return JSON.parse(configRevisionsRaw);
+};
+
+export class AutoblocksConfig<T> {
+  private _value: T;
+  private refreshIntervalTimer: NodeJS.Timer | undefined;
+
+  constructor(value: T) {
+    this._value = value;
+  }
+
+  private makeRequestUrl(args: { id: string; version: ConfigVersion }): string {
+    const configId = encodeURIComponent(args.id);
+    let version: string;
+    if (args.version === ConfigSpecialVersion.DANGEROUSLY_USE_UNDEPLOYED) {
+      version = 'undeployed';
+    } else {
+      version = encodeURIComponent(args.version);
+    }
+
+    return `${API_ENDPOINT}/configs/${configId}/${version}`;
+  }
+
+  private async getConfig(args: {
+    id: string;
+    version: ConfigVersion;
+    timeoutMs: number;
+    apiKey: string;
+  }): Promise<Config | undefined> {
+    const url = this.makeRequestUrl({ id: args.id, version: args.version });
+
+    const resp = await fetch(url, {
+      method: 'GET',
+      headers: {
+        ...AUTOBLOCKS_HEADERS,
+        Authorization: `Bearer ${args.apiKey}`,
+      },
+      signal: AbortSignal.timeout(args.timeoutMs),
+    });
+    const data = await resp.json();
+    return zConfigSchema.parse(data);
+  }
+
+  private async loadAndSetRemoteConfig(args: {
+    id: string;
+    version: ConfigVersion;
+    apiKey: string;
+    timeout?: TimeDelta;
+    parser?: (config: unknown) => T | undefined;
+  }) {
+    const remoteConfig = await this.getConfig({
+      id: args.id,
+      version: args.version,
+      timeoutMs: convertTimeDeltaToMilliSeconds(
+        args.timeout || { seconds: 10 },
+      ),
+      apiKey: args.apiKey,
+    });
+
+    if (!remoteConfig) {
```

Comment:
i think can drop assuming the return type of `getConfig` is `Config` and not `Config | undefined`